File: 29Items-Projects/7Algorithms/7-String-algorithms/src/main/java/com/stringalgorithms/algorithms/Trie.java
package com.stringalgorithms.algorithms;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public final class Trie {
    private final Node root = new Node();

    public void insert(String word) {
        Objects.requireNonNull(word, "word must not be null");
        Node current = root;
        for (char ch : word.toCharArray()) {
            current = current.children.computeIfAbsent(ch, ignored -> new Node());
        }
        current.word = true;
    }

    public boolean contains(String word) {
        Node node = findNode(word);
        return node != null && node.word;
    }

    public boolean startsWith(String prefix) {
        return findNode(prefix) != null;
    }

    private Node findNode(String value) {
        Objects.requireNonNull(value, "value must not be null");
        Node current = root;
        for (char ch : value.toCharArray()) {
            current = current.children.get(ch);
            if (current == null) {
                return null;
            }
        }
        return current;
    }

    private static final class Node {
        private final Map<Character, Node> children = new HashMap<>();
        private boolean word;
    }
}

File: 29Items-Projects/7Algorithms/7-String-algorithms/python/string_algorithms/algorithms/trie.py
from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class _Node:
    children: dict[str, "_Node"] = field(default_factory=dict)
    word: bool = False


class Trie:
    def __init__(self) -> None:
        self._root = _Node()

    def insert(self, word: str) -> None:
        current = self._root
        for char in word:
            current = current.children.setdefault(char, _Node())
        current.word = True

    def contains(self, word: str) -> bool:
        node = self._find_node(word)
        return node is not None and node.word

    def starts_with(self, prefix: str) -> bool:
        return self._find_node(prefix) is not None

    def _find_node(self, value: str) -> _Node | None:
        current = self._root
        for char in value:
            current = current.children.get(char)
            if current is None:
                return None
        return current
